src/ccs.rs

```diff
@@ -0,0 +1,636 @@
+  // Vector W in F^{n - l - 1}
+  W: Vec<G::Scalar>,
+}
+
+// TODO: Make sure this is in the right form for committed CCS using MLE, possibly a separate type
+/// A type that holds an CCS instance
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(bound = "")]
+pub struct CCSInstance<G: Group> {
+  // XXX: Move commitment out of CCSInstance for more clean conceptual separation?
+  // (Pedersen) Commitment to a witness
+  pub(crate) comm_W: Commitment<G>,
+
+  // Public input x in F^l
+  pub(crate) X: Vec<G::Scalar>,
+}
+
+/// A type that holds the shape of a Committed CCS (CCCS) instance
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(bound = "")]
+pub struct CCCSShape<G: Group> {
+  // Sequence of sparse MLE polynomials in s+s' variables M_MLE1, ..., M_MLEt
+  // TODO This should be MLE, but possible we don't have to keep in struct?
+  // Exists in paper but not multifolding-poc
+  pub(crate) M_MLE: Vec<SparseMatrix<G>>,
+
+  // XXX Embed CCS directly here or do a flat structure?
+  pub(crate) ccs: CCSShape<G>,
+  // q multisets S (same as CCS)
+  // q constants c (same as CCS)
+}
+
+/// CCCS Instance is (C, x)
+/// CCCS Witness is w _mle
+
+// NOTE: We deal with `r` parameter later in `nimfs.rs` when running `execute_sequence` with `ro_consts`
+/// A type that holds a CCCS instance
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(bound = "")]
+pub struct CCCSInstance<G: Group> {
+  // Commitment to a multilinear polynomial in s' - 1 variables
+  pub(crate) C: Commitment<G>,
+
+  // $x in F^l$
+  pub(crate) X: Vec<G::Scalar>,
+}
+
+/// A type that holds a witness for a given CCCS instance
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+pub struct CCCSWitness<G: Group> {
+  // Multilinear polynomial w_mle in s' - 1 variables
+  pub(crate) w_mle: Vec<G::Scalar>,
+}
+
+// NOTE: We deal with `r` parameter later in `nimfs.rs` when running `execute_sequence` with `ro_consts`
+/// A type that holds a LCCCS instance
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(bound = "")]
+pub struct LCCCSInstance<G: Group> {
+  pub(crate) C: Commitment<G>,
+  pub(crate) X: Vec<G::Scalar>,
+  pub(crate) u: G::Scalar,
+  pub(crate) v: Vec<G::Scalar>,
+}
+
+/// A type that holds a witness for a given LCCCS instance
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+pub struct LCCCSWitness<G: Group> {
+  pub(crate) w_mle: Vec<G::Scalar>,
+}
+
+impl<G: Group> CCS<G> {
+  // TODO: Update commitment_key variables here? This is currently based on R1CS with M length
+  /// Samples public parameters for the specified number of constraints and variables in an CCS
+  pub fn commitment_key(S: &CCSShape<G>) -> CommitmentKey<G> {
+    let num_cons = S.num_cons;
+    let num_vars = S.num_vars;
+    let total_nz = S.M.iter().fold(0, |acc, m| acc + m.0.len());
+
+    G::CE::setup(b"ck", max(max(num_cons, num_vars), total_nz))
+  }
+}
```

Comment:
I know this just follows the R1CS boilerplate. But we should really examine if we need this specific struct.